SVN: perform non-recursive revert

kevinushey opened the pull request:
This PR fixes an issue where attempts to 'revert' changes to a directory (ie, changes to its properties; for example, ignore properties) would cause a recursive revert, thereby reverting all changes files within that directory.

jjallaire commented:
My concern is that I put this in for some reason (i.e. it was not a day one thing but added to address some scenario):

https://github.com/rstudio/rstudio/commit/21b3e7a1fb203e723ff7d2a126d72a3d2e8d0631

@jcheng5 I don't want to give you any unpleasant flashbacks but do you have any recollection of why/when we'd want to do recursive reverts of directories in SVN.

@kevinushey We might need to play around a bit with various states to see if there is a real scenario where a recursive revert is needed.

kevinushey commented:
Here's a scenario for when you might want a recursive revert:

1. You create a directory,
2. You populate it with a number of files,
3. You add the directory,
4. You change your mind, and try to revert the directory.

You then get the error:

```
svn: E155038: Try 'svn revert --depth infinity' instead?
svn: E155038: Can't revert '/Users/kevin/scratch/R/dir' without reverting children
```

It seems like we need to somehow detect what kind of revert is actually appropriate; otherwise, we might need separate actions for 'Revert Properties' and 'Revert' or something to that effect.

kevinushey commented:
Alright, I've updated the PR: we now attempt to infer whether a file should be reverted with a recursive vs. a non-recursive revert.

We do this by taking the set of input paths, separating them into two separate vectors (those requiring recursive reverts vs. those not requiring recursive reverts), and then executing two separate `svn revert` invocations.

Right now, regular files, and newly-added directories (not modified directories) will get recursive reverts; modified directories (e.g. properties changed) get non-recursive reverts. This seems in line with the expected behavior.

@jjallaire, do you know if it's okay to re-use a single `ProcessResult` object with two separate commands? (Does the `stdout` and `stderr` just get appended, or should we use two separate process results and join them at the end?)

jjallaire commented on src/cpp/session/modules/SessionSVN.cpp:
```diff
@@ -733,13 +736,73 @@ Error svnRevert(const json::JsonRpcRequest& request,
+         const source_control::FileWithStatus& fileStatus =
+               fileStatusMap[key];
+         
+         shouldRevertRecursively = fileStatus.status.status() == "A";
```
Is there an analogous state for reverting the deletion of an entire directory?

kevinushey replied:
There is, but it appears a non-recursive revert is sufficient for reverting its deletion (alongside any files that lived within that directory).

jjallaire replied:
Okay, great!

On Wed, Mar 16, 2016 at 4:29 PM, Kevin Ushey <notifications@github.com>
wrote:

> In src/cpp/session/modules/SessionSVN.cpp
> <https://github.com/rstudio/rstudio/pull/637#discussion_r56408622>:
>
> > +   BOOST_FOREACH(const FilePath& filePath, paths)
> > +   {
> > +      if (!filePath.isDirectory())
> > +      {
> > +         recursiveReverts.push_back(filePath);
> > +         continue;
> > +      }
> > +
> > +      bool shouldRevertRecursively = false;
> > +      std::string key = filePath.absolutePath();
> > +      if (fileStatusMap.count(key))
> > +      {
> > +         const source_control::FileWithStatus& fileStatus =
> > +               fileStatusMap[key];
> > +
> > +         shouldRevertRecursively = fileStatus.status.status() == "A";
>
> There is, but it appears a non-recursive revert is sufficient for
> reverting its deletion (alongside any files that lived within that
> directory).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/637/files/8706f880fd73c8850554a4009fd6012228cbc1da#r56408622>
>

jjallaire commented on src/cpp/session/modules/SessionSVN.cpp:
```diff
@@ -733,13 +736,73 @@ Error svnRevert(const json::JsonRpcRequest& request,
+                     "--depth" << "infinity" <<
+                     "--" << recursiveReverts,
+                     true,
+                     &result);
```
Output will be appended (which is fine). The exitStatus will be just the status of the second command so you are losing the exit status of the first command (which may be okay here).

jjallaire commented:
Let's add a NEWS item for this.
